[deployment] prefix support for multi-tenancy

brettswift opened the pull request:
Intended to solve #195 

This adds functionality to the **prefix** config variable, changing the type from _boolean_ to a  _boolean/String_ hybrid. 

This feature is backwards compatible - existing deployments of r10k should be unaffected, as true/false are still valid, and the default value is still false.

adrienthebo commented on lib/r10k/source/base.rb:
```diff
@@ -23,16 +23,27 @@ class R10K::Source::Base
     @name    = name
     @basedir = basedir
-    @prefix  = options.delete(:prefix)
+    @prefix  = derive_prefix(name,options.delete(:prefix))
```
style: should be `...name, options.delete...`

adrienthebo commented on lib/r10k/source/base.rb:
```diff
@@ -23,16 +23,27 @@ class R10K::Source::Base
   end
 
+  def derive_prefix(name,prefix)
+    if prefix == true || prefix == 'true'
```
The special handling of `'true':String` is inconsistent with the documentation.

In the case where someone had a truthy value that wasn't true this would change the behavior, but I don't think that case is worth worrying about.

brettswift replied:
Corrected documentation.

I see what you mean about truthy values.   The fix here would be to check 

`if (prefix.is_a?(TrueClass) && prefix)` 

 What's the best practice here?   I don't mind changing it, let me know what is best.

elyscape replied:
You could just use:

``` ruby
if prefix == true
```

Truthy values aren't coerced into booleans, so this will only succeed if `prefix` is actually `true`.

adrienthebo commented on spec/unit/source/git_spec.rb:
```diff
@@ -139,6 +139,36 @@
   end
 end
+#
+# describe R10K::Source::Git, 'when prefixing is enabled, and the prefix name is overridden' do
```
These tests need to be implemented or deleted, committing in commented out code doesn't provide any value.

brettswift replied:
oops.  my bad.   I thought I reviewed my stuff well enough to catch this.

adrienthebo commented on lib/r10k/source/base.rb:
```diff
@@ -23,16 +23,27 @@ class R10K::Source::Base
   #   semantics of this hash may depend on the source implementation.
   #
-  # @option options [Boolean] :prefix Whether to prefix the source name to the
+  # @option options [Boolean | String] :prefix Whether to prefix the source name to the
```
style: multiple return values should be comma separated.

adrienthebo replied:
This comment is now out of date; it doesn't specify the behavior of a string prefix.

adrienthebo commented on spec/unit/source/git_spec.rb:
```diff
@@ -157,9 +187,9 @@
-    it "prepends the source name when prefixing is on" do
-      bn = described_class.new('mybranch', {:prefix => true, :sourcename => 'foo'})
-      expect(bn.dirname).to eq 'foo_mybranch'
+    it "prepends the prefix name when prefixing is overridden" do
```
This should be a new test case and should not replace the existing test case.

brettswift replied:
The flow here changed a bit.   `true` won't actually make it into the `BranchName` class anymore, due to the `derive_prefix` function in source/base.pp

https://github.com/brettswift/r10k/blob/feature/master/multi_tenant_prefix/lib/r10k/source/base.rb#L36-L44

So in BranchName, the only possible values are `false` or `:String`. 

I'm updating the documentation for `initialize` in this class to reflect that.

brettswift replied:
The tests that articulate this behaviour are found here: 

https://github.com/brettswift/r10k/blob/feature/master/multi_tenant_prefix/spec/unit/source/base_spec.rb#L33-L50

You'll see that the only output is false or string. 

It's a little bit unintuitive than my original fix which was done at the `source/git.pp`  but doing it at `source/base.pp` is a little more appropriate I think.

What do you think about this flow?

elyscape commented on doc/dynamic-environments/configuration.mkd:
````diff
@@ -121,6 +121,10 @@ sources:
 ```
+#### prefix behaviour
+* `true` prefix will be set to the `source` name.
+* `[string]` prefix will be set to the
````
This sentence is incomplete.

elyscape commented on lib/r10k/source/base.rb:
```diff
@@ -12,7 +12,7 @@ class R10K::Source::Base
 
   # @!attribute [r] prefix
-  #   @return [true, false] Whether the source name should be prefixed to each
+  #   @return [true, false, String] Whether the source name should be prefixed to each
```
Documentation does not specify what a `String` value would mean.  It might make more sense to change this to being either a `String` or `nil`, with the value being the prefix name.

brettswift replied:
Yea I agree now that you mention it.  A possible return val of false implies true might be possible.  I'm changing this to nil / String.

elyscape commented on lib/r10k/source/base.rb:
```diff
@@ -23,16 +23,27 @@ class R10K::Source::Base
   #   semantics of this hash may depend on the source implementation.
   #
-  # @option options [Boolean] :prefix Whether to prefix the source name to the
+  # @option options [Boolean, String] :prefix Whether to prefix the source name or prefix name to the
```
Unclear. Should probably read

> If a String, the prefix name for the environment directory names. Otherwise, whether to use the source name as the prefix. All sources should respect this option.

brettswift replied:
Yes, that was some brilliant Engrish on my part.   :)  Thanks for finding it.

elyscape commented on lib/r10k/source/git.rb:
```diff
@@ -44,7 +44,7 @@ class R10K::Source::Git < R10K::Source::Base
   #
   # @option options [Boolean] :prefix Whether to prefix the source name to the
-  #   environment directory names. Defaults to false.
+  #   environment directory names. Defaults to false. Will be a String of the prefix otherwise.
```
Should be brought in line with the change in base.rb and have the type changed to `[Boolean,String]`.  Also, as per [my comment on that change](https://github.com/puppetlabs/r10k/pull/295/files#r24205559), this would be clearer if it read:

> If a String, the prefix name for the environment directory names. Otherwise, whether to use the source name as the prefix. Defaults to false.

adrienthebo commented on lib/r10k/source/base.rb:
```diff
@@ -23,16 +23,27 @@ class R10K::Source::Base
     @options = options
   end
 
+  def derive_prefix(name,prefix)
```
This should be a private method.

adrienthebo commented on lib/r10k/source/base.rb:
```diff
@@ -65,4 +65,17 @@ def accept(visitor)
+
+  private
+
+  def derive_prefix(name,prefix)
```
So now that you've changed this to how I asked, I want you to change it again! Yay! :)

Per your comment on https://github.com/puppetlabs/r10k/pull/295/files#r24181790, the BranchName class will never be passed true, only false and a string. However the BranchName class is currently encapsulating most of the logic of naming branches, but this code touches branch naming as well. Instead of deriving the prefix here, why don't we push that into the BranchName class and just pass it through here? It'll help decouple things and prevent the Source classes from having to know too much about how prefixing is implemented.

In addition, both the Git and SVN sources have a BranchName method but I think there's no difference between them; it might be reasonable to extract the common code into `R10K::Source::BranchName` and deduplicate those classes.

adrienthebo replied:
For the record, in Source::Git there's one place where prefix is checked, but it's dead code; https://github.com/puppetlabs/r10k/pull/341 will remove that.

brettswift replied:
Ok I'll try to look at this soon.  Might find time tomorrow.

adrienthebo replied:
I thought I commented on this before but it doesn't look like I did; since this deals with branch name munging I would like this responsibility to be moved to the BranchName class so all the name manipulation is in a single place.

Given this change, the type of `@prefix` will be `[Boolean, String]`.

adrienthebo commented on doc/dynamic-environments/configuration.mkd:
````diff
@@ -121,6 +121,10 @@ sources:
     prefix: true # All environments will be prefixed with "mysource_"
 ```
+#### prefix behaviour
+* if `true` environment folder will be prefixed with the `source` name.
````
This reads as if it's literally "source" - might be good to rephrase to something like "the name of the source".

adrienthebo commented on doc/dynamic-environments/configuration.mkd:
```diff
@@ -121,6 +121,10 @@ sources:
+#### prefix behaviour
+* if `true` environment folder will be prefixed with the `source` name.
+* if `String` environment folder will be prefixed with the `prefix` value.
+
```
No mention of a false value, or default value.

adrienthebo commented on lib/r10k/source/git.rb:
```diff
@@ -43,8 +43,9 @@ class R10K::Source::Git < R10K::Source::Base
   #
-  # @option options [Boolean] :prefix Whether to prefix the source name to the
-  #   environment directory names. Defaults to false.
+  # @option options [Boolean, String] :prefix A String value to prefix the environment directory name.
```
This first line is misleading - best indicate that if it's a String, X occurs, else if it's a Boolean, Y occurs.

brettswift replied:
Ok.  I can't remember what I originally had,  I had changed this from previous feedback.  I'll clean it up. 

The same sentence is copied in these files: 
- environment/branchname.rb (after it's moved)
- source/base.rb
- source/git.rb  (err... uh.. i'll remove the docs from this class.    Honestly this is some of the first ruby code I've ever written so I'm not sure how the rdocs will behave but this might end up in duplicate documentation if it's also in the base class?)

brettswift replied:
uh, scratch that too :), I see the rest of the args are documented as well - i'll fix all three locations.

adrienthebo commented on lib/r10k/source/branchname.rb:
```diff
@@ -0,0 +1,112 @@
+# The name can be modified with a prefix, and by default will be automatically
+# modified if the branch name contains characters invalid for a file system.
+#
+class R10K::Source::BranchName
```
The `R10K::Source` namespace deals with collecting and creating environments, while BranchName is specific to a single environment. I would rather this live in the `R10K::Environment` namespace.

adrienthebo commented on lib/r10k/source/branchname.rb:
```diff
@@ -0,0 +1,112 @@
+require 'r10k/git'
```
None of these required files are used in this class.

adrienthebo commented on lib/r10k/source/branchname.rb:
```diff
@@ -0,0 +1,112 @@
+  # @!attribute [r] remote
+  #   @return [String] The url of the svn remote. Only used to storage and
+  #     retrieval when working with this class.
+  attr_reader :remote
```
This variable looks like it's coming from the `SVN::BranchName` class, but that class never used that value. We can omit it here.

adrienthebo replied:
Wait, that might be wrong on my part. Stay tuned.

brettswift replied:
Yea, when I first started refactoring I had the exact same thought process.  I nuked it, then it broke the svn class. 

I think I mentioned that in my comment above.   It could be made into a more generic hash, or SVN could handle things better and not need to push anything into BranchName just for storage. 

I let it go on first round of refactoring, not sure if SVN will be end of life at some point? : )

adrienthebo commented on spec/unit/source/base_spec.rb:
```diff
@@ -28,4 +28,5 @@
       subject.accept(visitor)
     end
   end
+
```
Spurious whitespace change

brettswift committed:
[deployment] prefix support

This adds behaviour to the configuration variable 'prefix' in
R10K::Source::Base. It allows that parameter to be a String to
declare the prefix name instead of just taking the source name
as the prefix name. This behaviour happens in
R10K::Environment::Name.
